fluka_sobp_source/source_sampler.f

```diff
@@ -0,0 +1,635 @@
+      END IF
+
+      END DO
+
+      RETURN
+      END
+
+
+!! ==================================================================================================
+!! ==================================================================================================
+
+!> @brief
+!! Reads beam configuration file
+!! @param[in] FILEPATH  path to the beam configuration file
+!! @param[out] ENERGY  particle energy in GeV/amu
+!! @param[out] DE  particle energy spread (sigma) in GeV/amu
+!! @param[out] XPOS beam spot center (X coordinate), in cm
+!! @param[out] YPOS beam spot center (Y coordinate), in cm
+!! @param[out] FWHMX beam spot size in X axis, in cm
+!! @param[out] FWHMY beam spot size in Y axis, in cm
+!! @param[out] PART beamlet weight
+!! @param[out] NOCOLUMNS number of columns in the file, negative if file missing or corrupted
+!! @param[out] NWEIGHT number of data rows in the file, negative if file missing or corrupted
+      SUBROUTINE READSOBP ( FILEPATH,
+     $   ENERGY, DE, XPOS, YPOS,FWHMX, FWHMY, PART,
+     $   NOCOLUMNS, NWEIGHT )
+
+      INCLUDE '(DBLPRC)'
+      INCLUDE '(DIMPAR)'
+      INCLUDE '(IOUNIT)'
+
+      CHARACTER*(*) FILEPATH   ! path to the sobp file
+      CHARACTER(8192) LINE
+      DOUBLE PRECISION ENERGY(65000), DELTAE(65000)
+      DOUBLE PRECISION XPOS(65000), YPOS(65000)
+      DOUBLE PRECISION FWHMX(65000), FWHMY(65000)
+      DOUBLE PRECISION PART(65000)
+      INTEGER NOCOLUMNS
+      LOGICAL LEXISTS
+
+      WRITE(LUNOUT,*) 'SOBP SOURCE READING ', FILEPATH
+
+*     warn if sobp.dat file is missing, stop calculation
+      INQUIRE(FILE=FILEPATH,EXIST=LEXISTS)
+      IF(.NOT. LEXISTS) THEN
+         WRITE(LUNOUT,*) 'SOBP FILE sobp.dat missing'
+         RETURN
+      END IF
+*
+*     open sobp.dat for reading
+      OPEN(44, FILE=FILEPATH,STATUS='OLD')
+*
+*     we will now probe the file to get the number of columns with numbers
+*     we skip comment lines, first non-comment line will be saved to LINE
+*     later we rewind the file to the beginning and read it with proper column format
+      LINE(1:1) = '*'
+      DO WHILE( LINE(1:1) .EQ. '*' )
+         READ(44,'(A)',END=10) LINE
+      END DO
+      REWIND(44)
+*     WRITE(LUNOUT,*) 'SOBP FIRST LINE', TRIM(LINE)
+*
+*     get number of columns from the first non-comment line
+      NOCOLUMNS = NOCOLS(LINE)
+      WRITE(LUNOUT,*) 'SOBP NUMBER OF COLUMNS', NOCOLUMNS
+
+
+      DO
+*        fortran arrays start with 1, so we increase the counter as the loop starts
+         NWEIGHT = NWEIGHT + 1
+         IF (NWEIGHT .GT. 65000) THEN
+            WRITE(LUNOUT,*) 'SOBP SOURCE ERROR: too many beamlets'
+            RETURN
+         ENDIF
+
+*        skip comment lines, first non-comment line will be saved to LINE
+         LINE(1:1) = '*'
+         DO WHILE( LINE(1:1) .EQ. '*' )
+            READ(44,'(A)',END=10) LINE
+         END DO
+
+*        read the line, and guess the format from number of columns
+         IF( NOCOLUMNS .EQ. 5 ) THEN
```

Comment:
also, I would not use space after the paranthesis
in SH12A this would be: `IF (NOCOLUMS .EQ. 5) THEN`